Fix harpy singing

luringens opened the pull request:
## About the PR
Some changes to when my new favourite morphotype is allowed to sing:

- Muzzled: Stops singing, prevents singing
- Zombified: Stops singing, prevents singing
- Stunned: Stops singing
- Knocked down: Stops singing
- Sleeping: Stops singing, prevents singing
- Crit/Dead: Stops singing, prevents singing
- Glimmer mute event: Stops singing, prevents singing
- Enough damage to interrupt speech: Stops singing.
- Any time you can't speak, including mime's vow or mute characters: Prevents singing.

This should fix some of the bugs listed in #180.

## Why / Balance
N/A

## Technical details
HarpySingerSystem now does two sets of things to enforce these rules. First it subscribes to a lot of events to close the Instrument UI when a muzzle is equipped, they're knocked down, etc. Second, it subscribes to OpenUiActionEvent and explicitly goes before ActivatableUISystem in order to intercepts the MIDI UI opening action. This allows it to run a bunch of checks to see if the harpy should be allowed to sing, and if not, stop it from opening. It also warns the player when this happens.

## Media
![image](https://github.com/DeltaV-Station/Delta-v-rebase/assets/8277780/9af402f0-d981-410e-b122-80e613bff003)
![image](https://github.com/DeltaV-Station/Delta-v-rebase/assets/8277780/59440cdb-e16b-4bb8-8be7-76c47375774b)
- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
N/A

**Changelog**
:cl:
- fix: Harpies can now sing while cuffed, but not while muzzled, mute, zombified, or otherwise incapable of talking.
- add: You can now make a harpy stop singing by slipping, stunning, or hurting them.

luringens committed:
Stop singing when muzzled

luringens committed:
Changes to when harpies can sing

Stops harpies from opening the MIDI UI while muzzled. Also reworks how
harpies are stopped from singing when incapacitated, so that harpies are
allowed to sing while cuffed. And for UX sake, tells the player why they
can't sing when applicable. This should fix some of #180.

DebugOk reviewed (dismissed):
Very cool

luringens committed:
Refactor to remove SingerIncapacitatedEvent

luringens commented on Content.Server/DeltaV/Harpy/HarpySingerSystem.cs:
```diff
@@ -9,23 +16,91 @@ public sealed class HarpySingerSystem : EntitySystem
     }
 }
+
+public sealed partial class SingerMuzzledEvent : InstantActionEvent
```
I wish this wasn't necessary, but I couldn't find a way around it. ` SubscribeLocalEvent<HarpySingerComponent, GotEquippedEvent>` never got called, so this acts as an intermediary between the two.

luringens commented:
Alright, putting a checklist of things I need to add checks for and test. Let me know if anything should be added or removed from the list.
Conditions that should interrupt and/or prevent singing:
- [x] Gagged/muzzled
- [x] Incapacitated (crit or dead)
- [x] Zombified
- [x] Knocked down
- [x] Asleep
- [x] Muted trait
- [x] Psionic glimmer event mute

luringens committed:
Significantly clean up HarpySingerComponent

luringens committed:
Merge branch 'master' into fix/harpy-singing

VMSolidus commented:
The players apparently want to be able to play spooky zombie music and/or Thriller as zombie birds. I'm inclined to suggest having zombification change their instrument options to something cursed. Thats not a requirement, just a funny suggestion. Personally I lean towards zombie birds not being able to sing

luringens committed:
Do not allow singing while zombified, also refactoring

luringens commented:
@evilexecutive on a similar note, currently if you're typing in text chat and then you take damage from a set of categories and above a set amount, the game immediately appends something like `-OOF!` to the message and sends it automatically. I could reuse some of that code to stop singing when punched or otherwise taking a solid chunk of brute damage. This could be funny along with a sound effect, though for that we'd need an appropriate bird screech. Or just reuse a scream. Thoughts?

DebugOk commented:
> The players apparently want to be able to play spooky zombie music and/or Thriller as zombie birds. I'm inclined to suggest having zombification change their instrument options to something cursed. Thats not a requirement, just a funny suggestion. Personally I lean towards zombie birds not being able to sing

Zombie harpies shouldn't have the ability to sing anymore

luringens committed:
Additional singing conditions

Now stops singing if knocked down, stunned, or asleep. Also prevents
singing if the player for any reason can not talk.

luringens commented:
Current implementation:
- Muzzled: Stops singing, prevents singing
- Zombified: Stops singing, prevents singing
- Stunned: Stops singing
  - Lasts so short it'll take longer to reopen the window and pick a file than the effect lasts anyways. And for consistency, stun doesn't prevent talking or cause stutter either.
- Knocked down: Stops singing
  - As above. Also means that harpies that have lost their wheelchair can sing even if they can't stand up.
- Sleeping: Stops singing, prevents singing (through !CanSpeak)
- Crit/Dead: Stops singing, prevents singing (through !CanSpeak)
- !CanSpeak: Prevents singing.

Turns out CanSpeak is a thing. Anytime you can't speak in chat, you now can't sing. This covers crit, dead, asleep, or being mute, simplying the code quite a bit. This should mean you can't sing while glimmer muted, and that mimes can't sing without breaking their wov of silence. Whether or not the latter makes sense I guess depends on how that vow is interpreted.

VMSolidus commented:
The latter example of Mime harpies not being able to sing is absolutely correct. Mimes should not he making ANY noise at all, as its part of their act. Being able to sing or imitate sounds would ruin the act, so I will say this counts for intended behavior.

My last question regarding implementation, does the !CanSpeak also stop a song in the case of the mute glimmer event? If it does, I wholeheartedly approve of these changes, you are doing amazingly good work!

VMSolidus commented:
Also as an aside, since !CanSpeak applies to Mute characters, yes that is also an intended feature that Mute harpies cannot sing. Which its good that we're covering that base too.

luringens committed:
Stop singing when muted status effect is applied

luringens commented:
I've added a handler that should stop singing when a `"Muted"` status effect is applied. `PsionicCatGotYourTongueRule` should trigger this, but I could not figure out how to force a station event to test it.

Do you want harpies to stop singing when taking 10+ brute/burn damage, the same way `DamageForceSaySystem` does it?

VMSolidus commented:
Yea, I will also agree that taking a chunk of damage like that should interrupt them. If it interrupts speech, it should also interrupt singing. Im all for having more options for players to get birds to shut up. This would also add some interesting emergent gameplay options. Such as a harpy salvager singing so that their team knows they're okay, and if something stops them, their team would know they're in trouble. 

Design wise, this is their main mechanical positive. They have no other mechanical benefits to weigh against their downsides, so adding mechanical complexity to this feature helps distinguish them more.

luringens committed:
Damage interrupts harpies when singing

luringens committed:
Merge branch 'master' into fix/harpy-singing

luringens commented:
Alright! Singing is now interrupted by the same criteria that speech is interrupted. I have updated the changelog and PR description, I think this is ready for review again now.

DebugOk commented:
> I've added a handler that should stop singing when a `"Muted"` status effect is applied. `PsionicCatGotYourTongueRule` should trigger this, but I could not figure out how to force a station event to test it.
> 
> Do you want harpies to stop singing when taking 10+ brute/burn damage, the same way `DamageForceSaySystem` does it?

`addgamerule PsionicCatGotYourTongue`, but it's broken anyways

DebugOk commented on Content.Server/DeltaV/Harpy/HarpySingerSystem.cs:
```diff
@@ -1,31 +1,160 @@
+            args.Handled = cantSpeak || muzzled || zombified;
+
+            // Explain why the user can not sing. One message is enough.
+            if (zombified)
```
I'd prefer a switch statement over if else

luringens replied:
There isn't a single variable being switched on per-value, so the only way I can think of to do this would be to switch on a dummy value, which doesn't look pretty:
```c#
switch (true)
{
    case true when zombified:
        _popupSystem.PopupEntity(Loc.GetString("no-sing-while-zombified"), uid, uid, PopupType.Medium);
        break;
    case true when cantSpeak:
        _popupSystem.PopupEntity(Loc.GetString("no-sing-while-no-speak"), uid, uid, PopupType.Medium);
        break;
    case true when muzzled:
        _popupSystem.PopupEntity(Loc.GetString("no-sing-while-muzzled"), uid, uid, PopupType.Medium);
        break;
}
```

DebugOk commented on Resources/Locale/en-US/deltav/harpy/singer_system.ftl:
`can not` sounds better as `cannot` or `cant`

luringens replied:
Agreed.

luringens committed:
Rename cantSpeak to canNotSpeak

luringens committed:
Merge remote-tracking branch 'upstream/master' into fix/harpy-singing

luringens commented:
Rewrote (force-pushed) those two commits to fix the committer metadata, no change to prior history. Sorry about that.

luringens committed:
Remove most cant-sing explanations

luringens committed:
Merge branch 'master' into fix/harpy-singing

Signed-off-by: Bakke <luringens@protonmail.com>

DebugOk commented on Resources/Locale/en-US/deltav/harpy/singer_system.ftl:
```diff
@@ -0,0 +1 @@
+no-sing-while-no-speak = You can not sing right now.
```
```suggestion
no-sing-while-no-speak = You can't sing right now.
```

luringens replied:
It looks like the text for [when you're muted](https://github.com/DeltaV-Station/Delta-v/blob/71a868654c23049cbee680886bf19099d4d869dc/Resources/Locale/en-US/speech/speech-effects.ftl#L1C1-L1C42) is `speech-muted = You can't speak right now!`, so I could add an excalamation mark to the current if we want to be consistent with that.

DebugOk replied:
You could if you want to, otherwise I'll just mark this as resolved

luringens replied:
~~Let's stick to "can not" , I think it is clearer, particularly since the text isn't that big and I opted not to make it bold red.~~ Sorry, I'm tired and misread the change 😅

DebugOk replied:
"Can not" isn't consistent with the other localization in the game, so no. Adding an exclamation mark I dont particularly care about, but it should still be cant

luringens committed:
Change can-not-sing feedback message

Co-authored-by: Debug <49997488+DebugOk@users.noreply.github.com>
Signed-off-by: Bakke <luringens@protonmail.com>

luringens committed:
Merge branch 'master' into fix/harpy-singing

Signed-off-by: Bakke <luringens@protonmail.com>

luringens commented:
🎉
